lighthouse-core/test/computed/layout-shift-variants-test.js

```diff
@@ -0,0 +1,189 @@
+      const shiftEvents = [];
+      for (let i = 0; i < 30; i++) {
+        shiftEvents.push({
+          score: 0.125,
+          ts: (i + 0.5) * 1_000_000,
+        });
+      }
+      const trace = makeTrace(shiftEvents);
+
+      const variants = await LayoutShiftVariants.request(trace, context);
+      expect(variants).toEqual({
+        avgSessionGap5s: 3.75,
+        maxSessionGap1s: 3.75,
+        maxSessionGap1sLimit5s: 0.75,
+        maxSliding1s: 0.25,
+        maxSliding300ms: 0.125,
+      });
+    });
+
+    it('calculates from two clusters of layout shift events', async () => {
+      const shiftEvents = [
+        {score: 0.0625, ts: 1_000_000},
+        {score: 0.1250, ts: 1_200_000},
+
+        {score: 0.2500, ts: 10_000_000},
+        {score: 0.5000, ts: 10_200_000},
+      ];
+      const trace = makeTrace(shiftEvents);
+
+      const variants = await LayoutShiftVariants.request(trace, context);
+      expect(variants).toEqual({
+        avgSessionGap5s: 0.46875,
+        maxSessionGap1s: 0.75,
+        maxSessionGap1sLimit5s: 0.75,
+        maxSliding1s: 0.75,
+        maxSliding300ms: 0.75,
+      });
+    });
+
+    it('calculates from three clusters of layout shift events', async () => {
+      const shiftEvents = [
+        {score: 0.0625, ts: 1_000_000},
+        {score: 0.2500, ts: 1_200_000},
+        {score: 0.0625, ts: 1_250_000}, // Still in 300ms sliding window.
+        {score: 0.1250, ts: 2_200_000}, // Sliding windows excluding most of cluster.
+
+        {score: 0.0625, ts: 3_000_000}, // 1.8s gap > 1s but < 5s.
+        {score: 0.2500, ts: 3_400_000},
+        {score: 0.2500, ts: 4_000_000},
+
+        {score: 0.1250, ts: 10_000_000}, // > 5s gap
+        {score: 0.1250, ts: 10_400_000},
+        {score: 0.0625, ts: 10_680_000},
+      ];
+      const trace = makeTrace(shiftEvents);
+
+      const variants = await LayoutShiftVariants.request(trace, context);
+      expect(variants).toEqual({
+        avgSessionGap5s: 0.6875,
+        maxSessionGap1s: 1.0625,
+        maxSessionGap1sLimit5s: 1.0625,
+        maxSliding1s: 0.5625,
+        maxSliding300ms: 0.375,
+      });
+    });
+
+    it('calculates the same LS score from a teeny tiny extra small cluster of events', async () => {
+      const shiftEvents = [];
+      for (let i = 0; i < 30; i++) {
+        shiftEvents.push({
+          score: 0.125,
+          ts: 1_000_000 + i * 10_000,
+        });
+      }
+      const trace = makeTrace(shiftEvents);
+
+      const variants = await LayoutShiftVariants.request(trace, context);
+      expect(variants).toEqual({
+        avgSessionGap5s: 3.75,
+        maxSessionGap1s: 3.75,
+        maxSessionGap1sLimit5s: 3.75,
+        maxSliding1s: 3.75,
+        maxSliding300ms: 3.75,
+      });
+    });
+
+    it('includes events with recent input at beginning of trace, but ignores others', async () => {
+      const shiftEvents = [
+        {score: 0.0625, ts: 250_000, had_recent_input: true},
```

Comment:
the complex mix of events and scores here made this hard for me to evaluate if it's satisfying the title here. WDYT about `score: 1, had_recent_input: true` for ~5 events, then 1 event 5s later with false then another cluster with true?

Reply:
> the complex mix of events and scores here made this hard for me to evaluate if it's satisfying the title here. WDYT about `score: 1, had_recent_input: true` for ~5 events, then 1 event 5s later with false then another cluster with true?

yeah, the test started as something else and then morphed into this, so that makes sense. SG